### PR TITLE
Add Tuttifrutti.DevelMain

### PR DESCRIPTION
**Pull request by zudov**
#6

### CODE REVIEW

**Comment by f-f** [comment]
@zudov I'll try to use this first in Persona, so we can evaluate the UX (as happened with #7)

File: src/Tuttifrutti/DevelMain.hs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    kill = do
+      Rapid.rapid rapidId $ \r -> do
+        envRef :: IORef env <- Rapid.createRef r ("env" :: String) $ do
+          error "Trying to close non-existing Env"
+        destroyEnv =<< readIORef envRef
+        Rapid.deleteRef r "env"
+        for_ threads $ \(name, _work) ->
+          Rapid.stop r name ()
+
+    withDevEnv :: forall a. (env -> IO a) -> IO a
```
Comment (f-f) [inline src/Tuttifrutti/DevelMain.hs]: Maybe I'd like this to be `runDev`

**Comment by zudov** [comment]
I of course already tried using it in Faro, so not much worries about UX, but of course try it in persona first

File: src/Tuttifrutti/DevelMain.hs (hunk 1 above)
Comment (zudov) [reply src/Tuttifrutti/DevelMain.hs]: There you go. I've defined `runRioDev` on top of `withDevEnv`.

**Comment by zudov** [comment]
Here's how I integrate in Faro:

```haskell
devHandle :: Dev.Handle Env
devHandle = Dev.newHandle 0 Dev.Config
  { Dev.createEnv = Env.newDevEnv
  , Dev.destroyEnv = Env.closeEnv
  , Dev.threads =
      [ ("api", \env -> do
            let port = 8080
            with env $ do
              Log.logInfo "reloading the dev web server" [ "port" .= port ]
            Main.faroApi env port
        )
      ]
  }

-- | Create 'Env' that persists across reloads and start
--   a webserver thread that uses it.
--
--   Running update repeatedly would reload the code in
--   the webserver thread, but won't recreate the 'Env'.
update :: IO ()
update = Dev.update develMain

-- | Destroy the 'Env' and kill the webserver thread.
--
--   Use this function when you want to do full restart.
kill :: IO ()
kill = Dev.kill develMain

runFaroDev :: Faro a -> IO a
runFaroDev = runRioDev develMain . RIO
```

**Comment by f-f** [comment]
This is how I'd like to integrate it in Persona:

```haskell
Dev.Module{..} = Dev.newDevelMain 0 Dev.Config{..}
  where
    createEnv = Env.newDevEnv
    destroyEnv = Env.closeEnv
    threads = [("webserver", \env -> Main.personaApi env 8080)]

```

**Comment by zudov** [comment]
This is how I first integrated it in Faro, but I decided against binding with RecordWildCards in this case. Mostly I didn't like the look of it (with type signatures). But it works and you can integrate in this way finely. I don't have a strong opinion about this, it just saves 3 lines of code.

**Comment by zudov** [comment]
On the other hand, I think we may disable some warnings in the DevelMain. I'm mostly thinking of unused imports (that are useful) and these type signatures.

But without type signatures and with wildcard bind there's no place to put the docstrings, no big deal though as it's a pattern that we document in the README and use everywhere.

**Comment by zudov** [comment]
In any case, I think I'd prefer the `runDev` to be a standalone definition like this (with Wildcards):

```haskell
devHandle@Dev.Handle{..} = Dev.newDevelMain 0 Dev.Config{..}

runDev :: RIO Env a -> IO a
runDev = runRioDev
```

Or like that (with no Wildcards):

```haskell
devHandle = Dev.newDevelMain 0 Dev.Config{..}

runDev :: RIO Env a -> IO a
runDev = runRioDev devHandle
```

**Comment by zudov** [comment]
Warnings can be disabled with `{-# OPTIONS_GHC #-}` pragma and some of these flags https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/using-warnings.html

**Pull request merged**